### ISSUE TITLE
Auto Differentiation, complex gradient,  is it expected?

### ISSUE THREAD

**Issue by carlos-luque** [qiboteam/qibo#1253]
I have attempted to run a test based on the test_variable_backpropagation to verify if the gradient becomes complex when the tf.variable is of type complex128, as indicated by the comment on line 897."

I ran the code:

```
from qibo.backends import construct_backend
backend = construct_backend("tensorflow")
import tensorflow as tf


theta = tf.Variable(0.1234, dtype=tf.complex128)

with tf.GradientTape() as tape:
    tape.watch(theta)
    c = Circuit(1)
    c.add(gates.X(0))
    c.add(gates.RZ(0, theta))
    result = backend.execute_circuit(c) 
    loss = tf.math.real(result.state()[-1])
    print ("state:")
    print(result.state())
grad = tape.gradient(loss, theta)
print("gradient:")
print(grad)
```

This produced:
```
state:
tf.Tensor([0.        +0.j         0.99809716+0.06166086j], shape=(2,), dtype=complex128)
gradient
tf.Tensor((-0.030830429966003096-0.4990485793869284j), shape=(), dtype=complex128)
```
Does the appearance of a complex gradient imply a modification in the target function? Is this an anticipated outcome, or does it suggest a potential issue?

**Comment by MatteoRobbiati** [comment]
Thanks for opening this. 
What you point out here is true, if we let TF free to explore complex numbers, it does, but to use complex angles in the rotations is indeed not correct. 

This has to be fixed, thanks again!

**Comment by renatomello** [comment]
> I remember in the past me and you fixed an issue related to TensorFlow warnings @renatomello, can you confirm? Those were warnings about we were discarding the imaginary part of the gradients (which is indeed non zero), but in the same time we were forcing the parameters to be `float64`.

The warnings were appearing in the examples in the documentation, but now I don't remember what they were related to.
https://github.com/qiboteam/qibo/pull/905

**Comment by alecandido** [comment]
The original script was missing a couple of imports from `qibo`, i.e.:
```py
from qibo import Circuit, gates
```

However, the wrong outcome is expected, as @MatteoRobbiati explained. The variable should be forced as a `float64`.

Then we can check how to limit the warnings (but the warnings are less of an issue wrt the wrong behavior).

**Comment by MatteoRobbiati** [comment]
> Then we can check how to limit the warnings (but the warnings are less of an issue wrt the wrong behavior).

Yes, indeed. We absolutely have to fix some examples in the documentation and, at the first stage, we can mute those warnings.
